expression/src/main/java/io/smallrye/common/expression/Expression.java

```diff
@@ -407,18 +410,47 @@ private static Node parseString(Itr itr, final boolean allowExpr, final boolean
                         }
                         case '$': {
                             // $$
-                            if (flags.contains(Flag.MINI_EXPRS)) {
-                                // TP 13
-                                list.add(new ExpressionNode(false, LiteralNode.DOLLAR, Node.NULL));
+                            if (flags.contains(Flag.NO_$$)) {
+                                if (itr.hasNext()) {
+                                    switch (itr.peekNext()) {
+                                        case '{':
+                                            list.add(new LiteralNode(itr.getStr(), start, idx));
+                                            break;
+                                        case '$':
+                                            list.add(new LiteralNode(itr.getStr(), start, idx));
+                                            itr.prev();
+                                            break;
+                                        default:
+                                            list.add(new LiteralNode(itr.getStr(), start, itr.getNextIdx()));
+                                            break;
+                                    }
+                                } else {
+                                    list.add(new LiteralNode(itr.getStr(), start, itr.getNextIdx()));
+                                }
                             } else {
```

Comment:
Now that the semantics of the flag are being more clear, couldn't we instead add one extra `if`/`else` to the next part (between TP 13 and TP 14) where if `NO_$$` is set we enqueue a `$` and re-enter the "got `$`" state, and then just forget everything else?